[2단계 - 점심 뭐 먹지] 캉골(김문경) 미션 제출합니다.

mun-kyeong opened the pull request:
## 학습 목표

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

- 어플리케이션을 컴포넌트 단위로 모듈화하여 개발
- UI를 컴포넌트 단위로 생각하고 개발하는 연습
- 재사용할 수 있는 컴포넌트를 고민해보기
- 웹 UI 환경에서의 테스트 기초
- 사용자 관점에서 중요하다고 생각하는 기능을 스스로 정의하고 E2E 테스트로 검증해보기


## 제출 전 체크 리스트

<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: https://mun-kyeong.github.io/javascript-lunch/

## 리뷰 요청 & 논의하고 싶은 내용

<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? -->

1. **domain과 ui를 분리하기 위해 신경을 많이 썼습니다.**

컴포넌트는 오직 컴포넌트로(UI)써의 역할만 수행할 수 있도록 구현했으며, 이외의 다른 기능들은 domain 로직으로 넘겨 수행할 수 있도록 했습니다. 

---

2. **localStorage에서 직접 가져오고 값을 변경하는 로직은 오직 `FoodItemHandler` 에서만 수행할 수 있도록 했습니다.**

localStorate에 접근하여 값을 가져오고 변경시키는 부분은 `FoodStorageHandler.ts` 에서 수행중입니다. 다른 곳에서 localStorage의 값을 가져오고싶거나 변경하고 싶을 땐 반드시 `FoodItemHandler`를 거쳐서 가져올 수 있도록 했습니다.

---

3. **filter 기능을 계속해서 유지한 채로 사용자에게 보여줄 수 있도록 했습니다.**

새로운 음식점을 추가하거나 삭제하거나 즐겨찾기를 선택할때나 해제할때나 모두 filter의 값을 인자로 전달받아서 필터링의 상태를 유지한 채로 화면을 보여줄 수 있도록 했습니다. 사용자에게 화면의 흐름이 부자연스럽지 않게 흘러갈 수 있도록 신경썼습니다.  

<br/>


### 2) 이번 리뷰를 통해 논의하고 싶은 부분

<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->


1. **모달창에서 즐겨찾기 버튼을 눌렀을 때의 action이 궁금합니다**

현재 구현은 모달창에서 즐겨찾기 버튼을 선택/해제 한 후에 모달창을 닫았을 때 결과값이 반영되도록 하였습니다. 모달창에서 즐겨찾기 버튼을 눌렀을 때 뒤에서 바로 값이 반영 되는 것은 계속해서 값을 렌더링 해야하기 때문에 자원이 낭비된다고 생각이 되었습니다. 또한 모달창에서 값을 수정하는 도중 모달창 뒤에서 어떠한 action이 일어나는 것은 부자연스러운 흐름이라 생각이 되었습니다.

> Q1 ) "현재 구현은 모달창에서 즐겨찾기 버튼을 선택/해제 한 후에 모달창을 닫았을 때 결과값이 반영되도록 하였습니다." 라는 로직이 부자연스럽지는 않은지 궁금합니다

---


2. **UI와 domain의 역할 분리 문제입니다.**

화면에 보여줄 껍데기 형식의 컴포넌트는 UI 라는 확신이 들지만 onClick 이나 버튼을 클릭했을 때 어떠한 event를 주는 것 또한 UI 로직으로 봐야 하는지 많은 고민이 되었습니다. 만약 onClick 속성 또한 domain 으로 본다면 컴포넌트와 onClick 속성을 가지는 또다른 class를 만들어야 하기에 이미 존재하는 컴포넌트에서 click 속성을 붙이기 위해 컴포넌트를 새로 만든 다는 것은 불필요하다고 느껴졌습니다. 

> Q2 ) 따라서 현재는 UI 로직에 onClick 속성과 handler를 전달하는 방식을 사용했는데 두 개가 분리되어야 할지 고민이 됩니다.

---


3. **e2e테스트에서 부족한 항목이 있을까요?**



 
---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계

- [ ] UI를 컴포넌트 단위로 분리했는가?
- [ ] 재사용 가능한 컴포넌트를 고려했는가?
- [ ] 주요 기능 시나리오에 대한 E2E 테스트를 작성했는가?

### 2단계

- [ ] 오류 상황(예: 잘못된 입력)에 대해서도 테스트 케이스를 정의했는가?
- [ ] 타입 정의를 적절히 사용하여 컴파일 오류 없이 안정적으로 동작하는가?
- [ ] 로컬/배포 환경에서 테스트가 정상적으로 동작하는지 확인했는가?

</details>

mun-kyeong committed:
[refactor] : FoodInventory domain 로직과 ui 로직 분리하기

mun-kyeong committed:
[feat] : 카테코리 목록을 확인할 수 있다 (ui)

mun-kyeong committed:
[feat] : 음식점 목록을 localStorage에 저장하고 불러올 수 있다.

mun-kyeong committed:
[feat] : 필터링 된 목록들을 확인할 수 있다.

mun-kyeong committed:
[teat] : 정렬된 목록들을 확인할 수 있다. (ui)

mun-kyeong committed:
[feat] : 카테고리, 이름, 거리, 설명, 참고 링크를 확인할 수 있다. (ui)

mun-kyeong committed:
[refactor] : localstorage로부터 도메인 로직 분리

mun-kyeong committed:
[feat] : 음식점을 삭제할 수 있다. (domain)

mun-kyeong committed:
[feat] : 즐겨찾기 버튼 + tabButton UI 작성

mun-kyeong committed:
[refactor] : foodDetail -> foodItem에서 flexDirection만 변경해서 사용가능하도록 변경

mun-kyeong committed:
[feat] : 즐겨찾기 버튼 click 가능하도록 수정

mun-kyeong committed:
[refactor] : 별 아이콘 위치 조정

mun-kyeong committed:
[feat] : 음식점 목록에서 자주 가는 음식점을 추가할 수 있다.

mun-kyeong committed:
[feat] : modal에서의 즐겨찾기 버튼 렌더링 방식 변경

mun-kyeong committed:
[refactor] : modal class에서 함수로 변경

mun-kyeong committed:
[refactor] : 컴포넌트 / 도메인 typescript 전환

mun-kyeong committed:
[refactor] : TS 적용 마무리

mun-kyeong committed:
[refactor] : page TS 마이그레이션

mun-kyeong committed:
[feat] : form 입력창 초기화 추가

mun-kyeong committed:
[style] : overflow 속성추가

mun-kyeong committed:
[refactor] : module 클래스 분리

mun-kyeong committed:
[refactor] : Tabbutton 컴포넌트 분리 및 코드 리팩토링

mun-kyeong committed:
[refactor] : 삭제 버튼 눌렀을 때 필터링 상태 그대로 렌더링 되도록 수정

mun-kyeong committed:
[refactor] : filter 속성에서 null 제거

mun-kyeong committed:
[refactor] : Modal 닫힐 시 리렌더링 되도록 수정

mun-kyeong committed:
[test] : e2e 테스트 추가

mun-kyeong committed:
[test] : e2e 테스트 추가

woowa-euijinkk commented on src/pages/FoodListPage.ts:
```diff
@@ -0,0 +1,102 @@
+    });
+
+    return new FoodListPage({ title, iconButton: AddFoodItemIcon });
+  }
```
왜 constructor 에서 실행하지 않고, static method 를 별도로 만드셨나요?
그냥 constructor 에서 실행하면 안 되나요?

mun-kyeong replied:
넵 맞는 것 같아요. class가 계속해서 사용되는게 아니라 맨 처음 페이지의 스켈레톤 코드가 로딩 될 때 최초의 한번만 실행되는 페이지여서 static을 이용하면서까지 class로 만들 필요가 없을 것 같습니다.
**한번만 실행하면 된다**라는 관점에서 해당 부분을 class가 아닌 일반 function으로 수정하게 되었습니다. 

[[refactor] : 폴더구조 변경](https://github.com/woowacourse/javascript-lunch/pull/240/commits/58db3ece4dd232f34c70a8f3417a958cd81523a0)

woowa-euijinkk replied:
넵 좋아요.

다만, 한번만 실행된다고 하여 무조건 function 일 필요는 없답니다.
다른 컴포넌트와의 일관성이 중요하다고 생각해요.
page 도 하나의 컴포넌트로 정의해보았을 때, 다른 컴포넌트도 class 도 선언했다면 이것도 class 인 편이 일관성 측면에서 좋겠어요.
일관성이 정말 중요합니다. 아무리 깔끔한 코드라도 일관적이지 않다면 읽기가 어려워집니다

woowa-euijinkk commented on src/pages/FoodListPage.ts:
```diff
@@ -0,0 +1,102 @@
+      this.#body?.appendChild(Header({ title, icon: iconButton }));
+    } else {
+      this.#body?.appendChild(Header({ title, icon: null }));
+    }
```
분기가 필요한가용?
이렇게만 있어도 되지 않을까요?

```js
this.#body?.appendChild(Header({ title, icon: iconButton }));
```

mun-kyeong replied:
넵 그렇더라구요🥹 어차피 `Header` 안에서 null 일때의 경우를 처리해주고 있어서 말씀하신 코드 방향으로 수정했습니다!

[[refactor] : 폴더구조 변경](https://github.com/woowacourse/javascript-lunch/pull/240/commits/58db3ece4dd232f34c70a8f3417a958cd81523a0)
+) 추가 설명 : 해당 부분을 class에서 function으로 수정하는 과정에서 함께 리팩토링 되었습니다!

woowa-euijinkk commented on src/pages/FoodListPage.ts:
```diff
@@ -0,0 +1,102 @@
+          </select>
+    `;
+    this.#main?.appendChild(container);
+  }
```
filter UI 가 중복이 있고, 재사용이 가능해 보이는데, 분리하면 어떨까요?

mun-kyeong replied:
그런 것 같아요. 반복되는 부분을 컴포넌트로 분리하니 훨씬 코드가 간결해진 것 같습니다! 좋은 의견 감사합니다.

[[refactor] : FoodListPage에서 반복되는 Filter 부분 컴포넌트화](https://github.com/woowacourse/javascript-lunch/pull/240/commits/a9df7b68560183ff5d516e91fc51db3f51b5b994)

woowa-euijinkk commented on src/domain/handler/TabButtonHandler.ts:
```diff
@@ -0,0 +1,22 @@
+      .querySelector(".tab-button .tab-button_favorite")
+      ?.classList.contains("selected-button") || false
+  );
+}
```
상태 -> UI 를 반영하는 구조여야지, UI 에서 상태를 가지고 오는게 이상해 보입니다. 이를 UI 가 "상태의 원천(source of truth)이다" 이라고 표현할 수 있는데요.

몇가지 문제가 예측돼요.

1. UI 변경 시 로직이 깨질 수 있습니다.
2. DOM 에 의존하기 때문에 테스트 하기 어렵습니다.
3. 상태를 single source of truth (SSOT) 로 관리하기 어렵습니다.

등이 있습니다.

상태와 UI 를 분리해보면 어떨까요?

woowa-euijinkk replied:
변수명도 이해가 어려워요. isFavoriteTabSelected , isFavoriteMode 등이 명확해보입니다!

mun-kyeong replied:
### 고민했던 부분
사실 처음에는 `상태 -> UI 를 반영하는 구조여야지, UI 에서 상태를 가지고 오는` 과정에 대해서는 크게 이상하다는 점을 느끼지 못했던 것 같아요. 현재 상태를 UI 에서 가져올 수도 있지 않을까? 하는 생각도 있었던 것 같습니다. 

다만 말씀하신대로 UI가 변경되었을 때 뿐 만 아니라 DOM에 의존되어서 테스트 하기 어렵다는 부분이 조금 많이 와닿았던 것 같아요. 이대로 흐린눈 하고 지나가기엔 언젠간 한번은 걸릴 것 같은 문제여서 조금 구현 방식을 고민을 많이 했던 것 같습니다.

---

### 변경점
현재 변경된 방식은 `FavoriteFilter` 라는 클래스를 사용하게 되었어요. 그리고 filter처럼 모든 곳에서 파라미터를 전달전달.. 받는 과정을 거치고 싶지가 않아서 구현 방식에 대해 고민이 조금 많았던 것 같습니다. **어떻게 하면 딱 한번만 FavoriteFilter를 생성**하고 이를 돌려쓸 수 있을까? 하는 고민이 되었던 것 같아요.

어쩌다가 주워들은 이야기로 class를 export하는게 아니라 class를 통해 생성된 객체 자체를 export 하라는 말이 떠올랐던 것 같아요. 그럼 해당 Class는 딱 한번만 실행되고 생성된 객체를 전역적으로 사용할 수 있으니까요.

---

###  결론 : 싱글톤 패턴 도입
그래서 이번에 해당 방법을 도입하던중 제가 사용하고자 했던 방식이 `싱글톤 패턴` 이라는 것을 알게 되었던 것 같습니다. 사실 아직 싱글톤 디자인 패턴에 대해서 구체적이게 알고 있는 것은 아니지만 인스턴스를 딱 한번만 실행한다는 관점에서 제가 사용하고자 했던 목적과 동일했던 것 같아요. 

조금 미숙하지만 해당 디자인패턴을 도입해서 말씀하셨던 `상태 -> UI 를 반영하는 구조여야지, UI 에서 상태를 가지고 오는` 과정을 수정해보았는데 혹시나 잘못 알고 사용한 부분이 있다면 정정 부탁드립니다!

[[refactor] : FavoriteFilter 싱글톤 배턴으로 변경](https://github.com/woowacourse/javascript-lunch/pull/240/commits/c0ade65751fe1224812048474fad8c2c8da463ff)

[[refactor] : Filter 싱글톤 배턴으로 변경](https://github.com/woowacourse/javascript-lunch/pull/240/commits/f3302d604211cbc39331b1dee18527ab1b56ab20)

woowa-euijinkk commented on src/domain/handler/FoodItemHandler.ts:
```diff
@@ -0,0 +1,100 @@
+import { handleTabButton } from "./TabButtonHandler";
+
+// CRUD - create : mock Data
+export function readFoodList({ favoriteFilter }: ReadFoodListType) {
```
`favoriteFilter` 이름이 이상합니다.
전체적으로 변수명이 직관적이지 않아, 단번에 코드를 이해하기가 어렵습니다! 
변수명을 더 잘 짓는 원칙들을 공부해보아요. 

코드를 볼 때, 세부적인 구현은 뒷전이고, 일단 전체적인 흐름만 봅니다. 변수명, 함수명, 인터페이스 정도로만 코드를 이해하고, 구현 자체는 필요 시에만 보죠. 그러니까, 이름만 잘 짓고 인터페이스만 잘 지어도, 전체 코드 흐름을 쉽게 이해할 수 있을거에요!

woowa-euijinkk commented on src/domain/handler/FoodItemHandler.ts:
```diff
@@ -0,0 +1,100 @@
+    localStorage.setItem("foodList", JSON.stringify(foodItems));
+  }
+  return previousFoodList;
+}
```
함수의 역할이 많습니다. 그리고 함수명이 그것을 담지 못 하고 있어요.

1. FoodList 를 받아와서, filtering 여부를 판단하고, filtering 한다.
2. 조건에 따라 기본값을 setItem 해준다. 

등이 보이네요. 한번 잘 분리해보아요!

woowa-euijinkk replied:
`previousFoodList` 도 이름이 왜 previous 가 붙는지 모르겠네용!

woowa-euijinkk commented on src/Module.ts:
```diff
@@ -0,0 +1,82 @@
+          const filteredItems = this.#filter.chageFilter({
+            foodList: previousFoodList,
+            filter: name as FilterType,
+          });
```
이렇게 바꾸면, 타입 단언 문제가 해결되겠어요. 
왜 그런지 tuple vs array 를 학습해보아요~!


```suggestion
    const filterTabs = ["category", "sorting"] as const
    filterTabs.forEach((name) => {
      document
        .querySelector(`select[name=${name}]`)
        ?.addEventListener("change", () => {
          const previousFoodList = readFoodList({
            favoriteFilter: isFavoriteState(),
          });

          const filteredItems = this.#filter.chageFilter({
            foodList: previousFoodList,
            filter: name,
          });
```

woowa-euijinkk commented on src/domain/Filter.ts:
```diff
@@ -0,0 +1,55 @@
+    if (filter === "category") this.#categoryFilter = filterOption;
+    else this.#sortingFilter = filterOption;
+    return this.updateFilterItem({ foodList });
+  }
```
changeFilter 가 foodListItem 을 return 하고 있는게 인터페이스가 이상합니다.
구조 변화가 필요하겠습니다

woowa-euijinkk commented on src/domain/Filter.ts:
```diff
@@ -0,0 +1,55 @@
+    const filteredItems = foodItems
+      ?.filter((foodItem) => {
+        if (this.#categoryFilter === "전체") return foodItem;
+        return foodItem.imgAlt === this.#categoryFilter;
```
imgAlt 랑 비교하고 있네요.
깨지기 쉬운 코드입니다!

woowa-euijinkk commented on src/domain/Filter.ts:
```diff
@@ -0,0 +1,55 @@
+    return this.updateFilterItem({ foodList });
+  }
+
+  updateFilterItem({ foodList }: UpdateFilterItemType) {
```
변수명이 이상합니다. 
updateFilterItem => FilterItem 이 뭐지? Filter 랑 Item 을 구분해서 봐야하나? Update 의 대상은 Filter 인가 Item 인가?

캉골 변수명에 신경을 많이 쓰셔야겠어요. 변수명 작성 훈련이 되어있지 않습니다. 그러다보니, 함수의 역할도 모호해지는 것 같습니다. 

저라면 `filterAndSortFoodList` 정도로 지을 것 같습니다. 그러면 명확히 함수의 역할을 이해할 수 있겠죠. 
이게 함수 역할이 2개라고 생각된다면, 함수를 2개로 다시 분리하면 되고요!

woowa-euijinkk commented on src/domain/Filter.ts:
```diff
@@ -0,0 +1,55 @@
+      return Number(a.distance) - Number(b.distance);
+    }
+    return 0;
+  }
```
상수를 관리해줍시다!

그리고 이 함수는 역할이 2개 입니다.
sortByKO, sortByNumber. 

분리하지 않아도 된다고 생각한다면, 왜 그런지 / 분리 해야 한다면 왜 해야하는지 생각해보아요.

woowa-euijinkk commented on src/Module.ts:
```diff
@@ -0,0 +1,82 @@
+import { FoodForm } from "./pages/FoodForm";
+import { FilterType } from "./types/domain/FilterType";
+
+export class Module {
```
이 클래스는 역할이 짬뽕이네요. 
선언을 편하게 하기 위해 묶어둔 클래스 같은데, 
module 의 이름이 "Module" 이라면, 이 module 의 역할은, "모든 module 을 관리한다" 이므로, 모든 module 이 사실상 여기에 들어올 수 있겠습니다.

이름도 잘못되었고, 이 기능들을 묶은 클래스가 필요한지, 안 필요한지, 고민이 필요합니다!

woowa-euijinkk commented on src/Module.ts:
```diff
@@ -0,0 +1,82 @@
+        this.#updateFoodList(key === "favorite");
+      });
+    });
+  }
```
함수 역할이 너무 많습니다. 어떻게 분리하면 좋을지 고민해보아요!

woowa-euijinkk commented on src/domain/handler/FoodItemHandler.ts:
```diff
@@ -0,0 +1,100 @@
+    return foodComponent;
+  });
+  showFoodItem({ foodListComponent: FoodItemListComponent || [] });
+}
```
이 부분 대대적으로 수정이 필요해보여요.

여러 문제가 있습니다.
filter 컴포넌트가, 여기저기 parameter 로 넘어가서 호출되고 있습니다.
convertStorageToLocal -> openDetailModal -> FoodDetail -> handleTabButton, deleteFoodItem 등 계속 넘어가고 있어요.
 Filter 내부 상태를 유지하기 위함이라면, Filter 를 어디 선언해두고, export/import 하면 될거같아요.

woowa-euijinkk commented on cypress/e2e/e2e.cy.js:
```diff
@@ -43,16 +43,94 @@ describe("E2E Test Group", () => {
+      });
+  });
+
+  it("'거리순' 필터링을 선택했을 때 음식점이 거리순으로 정렬되어 있는지 확인한다.", () => {
+    cy.get(".button-container button").eq(1).click();
+
+    cy.get("#sorting-filter").select("거리순");
+    cy.get(".restaurant-list .restaurant .restaurant__distance").then(
+      ($distances) => {
+        const distances = $distances
+          .toArray()
+          .map((el) => parseInt(el.innerText.match(/\d+/)?.[0] || "0"));
+        const sortedDistances = [...distances].sort((a, b) => a - b);
+        expect(distances).to.deep.equal(sortedDistances);
+      }
+    );
+  });
```
꼼꼼하게 테스트 추가해주셨네요. 👍

상수화, 함수화는 더 필요하겠어요!

woowa-euijinkk reviewed (changes requested):
캉골 안녕하세요!
2단계 구현하시느라 고생 많으셨어요.

전체적으로 일관된 문제가 많아서, 하나하나 리뷰하기 보다는, 어떤 것이 근본적인 문제일까 많이 고민해보았어요. 캉골의 근본적인 습관이 어떤 것이라서 지금의 구조가 나왔을까?  캉골이 어떤 것을 개선하면, 전체적인 코드 퀄리티가 확 좋아질까? 고민해보았어요. 
캉골의 성장을 위해서 피드백 드려요! 기분 나쁘게 들리지 않았으면 좋겠습니다. 🙏 

제가 판단하기론, 캉골의 좋지 않은 코드 습관이 몇 가지 있습니다. 

### 1. 변수명을 대충 짓는다.

코드를 빠르게 읽을 때, 인터페이스만 보고 코드를 읽습니다. 즉, 함수명, 변수명, 인자명만 보죠. 그렇게만 봐도, 어떤 흐름으로 어플리케이션이 돌아가는지 이해할 수 있어야 합니다. 하지만 캉골의 코드는 읽을때 중간에 턱턱 막힙니다. 사실 대부분의 코드에서 읽다가 막혀서, 세부 구현 자체를 살펴보아야 했습니다. 

그래서 캉골이 변수명을 대충 짓는 습관이 있다고 추론하게 되었어요.

그 이유는 다음과 같아요.
1. 말이 안되는 변수명이 있다.
`updateFilterItem`

2. 역할이 많아서, 이름을 모호하게 지어버린다.
`Module`, `~Handler` `handle~` `isFavoriteState` 

3. 역할이 많은데, 이름에 드러나지 않는다. 
`readFoodList` `addFoodFormItem` `convertStorageToLocal`  `chageFilter`

그러다보니, 함수 역할이 모호해집니다. 이름 자체가 모호하니, 함수에 모든 것을 담을 수 있게 되는것이죠. 

제가 캉골의 racingcar, lotto 의 코드를 살펴보지 않아서, 잘 모르겠으나, 이전 미션에서 잘 훈련이 되지 않았다고 느낍니다. 

### 2. 함수 분리가 잘 되지 않는다.
이건 아마 UI 단에 넘어오면서 발생한 문제 같습니다. 경계가 모호해지니, 함수 분리에 어려움을 겪으시는 것 같아요. 
구현을 할 때 분리를 하지 않고, 일단 구현을 하고, 분리한다는 생각으로 구현하시는 것 같아요.

함수를 먼저 쓰고 -> 구현을 하는 방식으로 코드를 짜면 역할이 더 명확해집니다. 
아직 완성되지 않은, 선언도 없는 함수를, 먼저 인터페이스만 호출하고, 그 다음에 내부 구현을 짜는거죠.
TDD 도 이와 같은 원리입니다. 기능을 먼저 생각하고 인터페이스를 먼저 짜는 습관은 전체 설계를 좋게 만드는데 도움을 줍니다.

### 3. 기존 코드를 유지한 채로 코드를 추가한다.
캉골의 코드를 보면, 왜 이렇게 까지 되었나 하는 포인트들이 있습니다. 
제가 그 원인을 타고타고 분석해보면, 그 시작점이 명확히 존재하더라고요. DOM 의 name 을 기준으로 생각한다거나, Filter 클래스를 여기저기 parameter로 넘겨주면서 코드를 진행하거나요. 

캉골도 아마 중간에 이렇게 해도 되나? 싶은 생각이 드셨으리라 생각합니다. 기존 구조에 얽매여서, 어쩔수 없이 이후 코드를 쓰고 있는 상황이요. 이게 맞나? 싶지만, tricky 하게 모두 해결은 가능하죠. 
그럴 때 한번 멈추고, 전체 구조를 다시 설계해서 개선시킬 필요가 있습니다. 그러지 않으면 그게 소위 말하는 legacy 가 되어버리죠. 
코드베이스가 작을 때부터 이런 습관을 가지는게 중요합니다.

캉골이 문제라고 생각하신 부분들을 나열해보고, 하나씩 리팩터링 해보는 경험을 가져보아요. 모두 할 필요는 없습니다. 하나의 성공 경험만이라도 가져가시면 좋겠어요!

---
그렇지만, 캉골이 구현력에서 큰 장점을 가진다고 느껴요. 
어떤 구조에서든, 어떤 방식으로든 구현을 해내니까요. 코드를 쓰는 것을 봐도 한줄한줄 로직을 잘 작성하시는 것 같아요. 
사실 이게 가장 중요하고, 이후에 코드 구조와 클린코드를 고민해보면 되는거죠. 

캉골에게 도움이 되는 서적을 추천드려요. `리팩터링` 책을 한번 읽어보면 도움이 많이 될 것 같아요.
변수명을 어떻게 지어야 하나, 인터페이스를 어떻게 해야 하나 등 읽기 쉬운 코드를 위한 내용이 많아요.

---

추가로 질문에 대해 답을 드릴게요.

### Q1 ) "현재 구현은 모달창에서 즐겨찾기 버튼을 선택/해제 한 후에 모달창을 닫았을 때 결과값이 반영되도록 하였습니다." 라는 로직이 부자연스럽지는 않은지 궁금합니다
오.. 자원의 낭비라고 생각하셨군요. 좋은 고민이네요.
- 성능 손실 vs 유지보수하기 좋은 코드

라는 대표적인 trade-off 관계인데요. 자주 고민해보시면 좋겠네요. 

1. 성능 손실이 얼마나 발생하는가?
이걸 측정해보셨나요? 왜 후자의 선택을 하셨나요? 어떻게 측정해보면 될까요?

2. 유지보수성은 생각보다 중요하다.
어느정도 성능 손실이 있더라도, 그게 가시적이지 않다면 유지보수를 선택하는게 좋습니다. 사실 일차적으로 유지보수를 선택하고, 성능 문제가 생기면 그때 성능을 택하고, 유지보수를 버리는 선택을 하는게 보편적인 방법 같네요. 

DOM 하나 바꾸는 비용이 얼마인지, 측정해보고, 다시 고민을 해보시죠!


### Q2 ) 따라서 현재는 UI 로직에 onClick 속성과 handler를 전달하는 방식을 사용했는데 두 개가 분리되어야 할지 고민이 됩니다.
이것도 좋은 질문인데요. 기본적으로 event handler 를 부착하는 함수는 UI 이고, 그 내부 요소는 UI 일수도 있고, domain 일 수도 있겠죠. 그게 hybrid 인 경우도 많고요.
event handler 가 controller 처럼, domain, UI 를 결합해서 호출하도록 많이 쓰는 것 같네요!

woowa-euijinkk commented:
캉골! 
모든 코멘트를 다 반영하실 필요 없고, 금하게 하실 필요도 없어요.

자신의 코드를 전체적으로 관망해보고, 다른 사람 코드들도 많이 읽어보고, 어떻게 짜면 좋을지, 앞으로 어떤 노력을 더 하면 좋을지 큰 그림 속에서 고민하는 것을 해보아요. 깨달음이나 방향을 공유해주셔도 좋고요.

마이크로한 문제는 일단 빼버리시죠!

mun-kyeong commented:
> 전체적으로 일관된 문제가 많아서, 하나하나 리뷰하기 보다는, 어떤 것이 근본적인 문제일까 많이 고민해보았어요. 캉골의 근본적인 습관이 어떤 것이라서 지금의 구조가 나왔을까? 캉골이 어떤 것을 개선하면, 전체적인 코드 퀄리티가 확 좋아질까? 고민해보았어요.

와 시지프 ㅠㅠ 정말 정확하게 봐주신 것 같아서 놀랐어요.. 사실 저도 코드를 작성하면서 계속 불편한 느낌이 드는데 이게 어디서부터 오는지를 잘 모르니 계속 답답한 채로 코드를 썼던 것 같아요. 처음에 `1. 변수명을 대충 짓는다.` 라는 문장을 봤을 때 부터 너무 핵심을 짚으셔서 웃음이 터졌어요 😂 은연중에 알고는 있었는지 몰라도 직접 마주하니 고쳐야 할 부분이라는 느낌이 많이 드네요. 

---

> 1. 변수명을 대충 짓는다.

너무 정확하게 짚어주신 부분입니다 🥹 그 함수가 어떤 역할을 하는지를 이름에 나타내가 위한 노력이 부족했던 것 같아요. 제공해주신 피드백처럼 우선 전반적으로 동작하는 기능을 만들기 위해 디테일한 부분에서 놓친게 많다는 생각이 들어요. 각 함수네이밍에서 조금 더 역할을 드러낼 수 있도록 노력해보겠습니다. 감사합니다.

<br/>

> 2. 함수 분리가 잘 되지 않는다.

`함수를 먼저 쓰고 -> 구현을 하는 방식으로 코드를 짜면 역할이 더 명확해집니다.`
오 이렇게 구현하는 부분에 대해선 미처 신경쓰지 못한 부분인 것 같아요. 말씀하신대로 TDD를 작성하는 방식과 비슷한 것 같아요. 이렇게 되면 함수 이름에서 먼저 역할이 드러나서 함수 네이밍에도 신경을 쓸 수 있을 것 같아요. 
`아직 완성되지 않은, 선언도 없는 함수를, 먼저 인터페이스만 호출하고, 그 다음에 내부 구현을 짜는거죠.` 
지금 미션에선 이미 구현을 먼저 해버린 상황이여서 함수 네이밍과 같은 문제가 벌어진 것 같아요. 다음 미션을 진행할땐 말씀하신 방법대로 구현해보는 것을 연습해보고 싶습니다! 

<br/>


> 3. 기존 코드를 유지한 채로 코드를 추가한다.

` DOM 의 name 을 기준으로 생각한다거나, Filter 클래스를 여기저기 parameter로 넘겨주면서 코드를 진행하거나요.`
맞아요 ㅠㅠ 처음에는 filter와 modal 을 모두 파라미터로 전달받다가 리팩토링을 거쳐 modal은 따로 함수로 뺄 수 있었어요  [[refactor] : modal class에서 함수로 변경](https://github.com/woowacourse/javascript-lunch/pull/240/commits/94a3ac9bab54067fd16f50743e500d062c3ecb68)

`그럴 때 한번 멈추고, 전체 구조를 다시 설계해서 개선시킬 필요가 있습니다.` 
제가 생각했을 때 (제게)부족한 부분도 레거시 코드가 되어버린 부분을 어느 한 순간에 멈추고 리팩토링을 거치는 시점을 잘 못잡는 것 같아요. **조금 불편하다**라는 느낌은 받지만 아직 기능이 구현이 되지 못한 상황에서 중간에 리팩토링을 거치는 것을 두려워하는 것 같습니다. 지금 제게 필요한 부분은 **불편하다**라는 느낌이 들었을 때 이를 **어떻게 하면 불편함을 해소할 수 있을까** 라는 고민을 할 시간이 필요하다고 생각이 듭니다. 

`코드베이스가 작을 때부터 이런 습관을 가지는게 중요합니다.`
말씀하신 대로 코드를 작성한 범위가 적을 때 수정하는 게 좋을 것 같아요. 앞으로 조금씩 이런 습관들을 고쳐나갈 필요가 있는 것 같아요 ㅠㅠ

<br/>

> 어떤 구조에서든, 어떤 방식으로든 구현을 해내니까요. 코드를 쓰는 것을 봐도 한줄한줄 로직을 잘 작성하시는 것 같아요.

시지프 최고.. 🥹 👍 제가 가진 장단점을 너무 잘 파악해주셔서 도움이 정말 많이 되는 것 같아요. 말씀하신 대로 **어떻게든 동작하는 기능**은 만들자. 라는 생각을 가지고 있어서 그런 부분은 강점으로 작용하지만 그만큼 제 약점이기도 한 것 같아요. 조금 더 디테일한 부분을 챙기면서 구현할 수 있도록 노력해보겠습니다.

<br/>

> Q1 ) "현재 구현은 모달창에서 즐겨찾기 버튼을 선택/해제 한 후에 모달창을 닫았을 때 결과값이 반영되도록 하였습니다." 라는 로직이 부자연스럽지는 않은지 궁금합니다

성능이 저하된다 라는 정확한 지표 없이 그저 지레짐작 한 부분입니다 🥹 성능이 저하된다고 생각했던 이유는 모달창에서 좋아요 버튼을 누를 때 마다 뒤에 페이지에서 계속해서 값을 렌더링 해야 했기 때문이예요. 더분다나 페이지를 로드 할 때 마다 필터링 값과 현재 페이지가 `자주 가는 음식점`인지 등을 확인해야 했기 때문에 cost가 많이 들 것 같다는 판단을 했습니다.

말씀하신대로 얼마나 성능에 영향을 미치는지 한번 확인을 해보고 싶어요. 리팩토링을 거치며 DOM 하나 바꾸는 비용이 얼마인지 한번 측정해보겠습니다! 

---

## 해결이 필요하다고 생각되는 부분

- [ ] filter를 파라미터로 계속해서 전달받고 있다.
- [ ] 이름이 모호한 함수들이 존재한다. 그 함수에서 기능들을 이름으로 나타내지 못하고 있다.
- [ ] 모달창에서 DOM을 바꾸는데 얼마의 비용이 드는지 측정해보기. +) 별표를 눌렀을 때 화면에 반영되는 나만의 기준을 세워보기

## 앞으로의 계획

- 리팩터링(마틴 파울러) 책 읽어보기
- TDD 방식과 유사하게 함수 껍데기를 먼저 작성해보고 구현을 해보기

mun-kyeong committed:
[refactor] : 기능 분리 및 이름 함수이름 리팩토링

mun-kyeong committed:
[refactor] : chageFilter 함수의 역할 분리

- 필터를 변경만 해주며 필터링된 배열을 받는 함수는 chageFilter 외부에서 진행하도록 수정

mun-kyeong committed:
[refactor] : updateFilterItem 명칭변경

- filterFoodList 으로 명칭변경. FoodList를 필터링한다는 의미가 담김

woowa-euijinkk commented:
캉골이 피드백을 어떻게 받아들이까, 걱정했는데요. 
제가 조금 강하게 이야기했음에도, 본인이 성장하는 기회로 받아들이시는 모습이 좋네요!
피드백을 받아들이고, 부족한 부분을 스스로 메타인지 하고, 개선할 수 있는 태도는 아주 귀중한 자세라고 생각합니다. 

저는 캉골이 더 성장할 수 있길 바라고, 그것을 도와줄 예정이니, 언제든 물어봐 주시고, 수용하기 어려운 부분이 있다면 말씀해주세요~!
2단계 수정 화이팅이에요! 💪 💪

mun-kyeong committed:
[refactor] : Module => AppState 으로 이름변경

mun-kyeong committed:
[refactor] : handleTabButton => changeFavoriteStatus 으로 이름변경

mun-kyeong commented:
> 피드백을 반영하며 수정된 부분들과 수정하게 된 이유들을 간략히 정리해보았습니다.

## 1. 변수명 수정

피드백을 반영하다보니 시지프가 언급한 `변수명에서 역할이 잘 드러나지 않는다` 라는 의미가 🚨**함수의 역할이 분리가 잘 되어있지 않다 라는 의미**🚨로 다가왔습니다. **예를들면 `chageFilter`의 경우엔 필터를 바꿀 뿐 만 아니라 필터링된 배열을 리턴 값으로 전달하는 역할까지 가지고 있더군요.** 그래서 함수의 이름을 바꾸면서 함수의 역할을 **이름**에서 나타내는 역할 까지만 가져올 수 있도록 수정하게 되었습니다.

변수명을 수정하게 된 예시와 그 이유를 간단히 정리해두겠습니다.

[[refactor] : 기능 분리 및 이름 함수이름 리팩토링](https://github.com/woowacourse/javascript-lunch/pull/240/commits/9e2396bdfbc411cf6b99b99a39d334b9e34853b0)

- `readFoodList` 
  - `getFilteredFoodList`로 명칭 변경. localStorage에서 값을 읽어오고 좋아요 표시가 된 item인지를 필터링해서 배열 반환

- `addFoodFormItem` 
  - `addFoodItem`으로 명칭변경. 인자값으로 foodItem을 받아서 다른 곳에서도 사용할 수 있도록 수정. `addFoodFormItem`이라는 이름은 FoodForm 페이지에서 onClick 로컬 함수로 변경해서 사용
  
 - `convertStorageToLocal`
   - `createFoodListComponent`으로 명칭 변경. 내부 함수에서 FoodItem 정보를 받아서 FoodItemComponent로 바꾸는 부분을 `createFoodItemComponent`이라는 함수로 분리해서 적용. 화면에 FoodListComponent을 생성해서 보여준다는 의미가 전달되도록 수정
   
[[refactor] : chageFilter 함수의 역할 분리](https://github.com/woowacourse/javascript-lunch/pull/240/commits/f9df8fbf19d756f011d4a440e9f785a431b3d477)

- `chageFilter` 
  - `chageFilter`가 필터를 변경할 뿐 만 아니라 필터링된 배열까지 제공하고 있었음. `chageFilter`는 필터를 변경만 하고 외부에서 Foodlist를 업데이트 할 수 있도록 수정

[[refactor] : updateFilterItem 명칭변경](https://github.com/woowacourse/javascript-lunch/pull/240/commits/15a53c383837d2e29e5ba2ba7dc83e55c8ca1bed)

- `updateFilterItem`
  - 필터링된 FoodList를 제공한다는 의미로 `filterFoodList`로 명칭 변경 
 
[[refactor] : Module => AppState 으로 이름변경](https://github.com/woowacourse/javascript-lunch/pull/240/commits/81b06b5cb030257d3b96fb371d1f706c7c4f0dcf)
[[refactor] : AppState => initAppState라는 함수로 수정](https://github.com/woowacourse/javascript-lunch/pull/240/commits/30c9456b0b0b773e3443395a027b4ed171d6ed21)

- `Module`
  - ~~AppState이라는 이름으로 변경. 초반에 Module이라는 이름을 지은 이유는 class, 특히나 객체를 생성할 때엔 어떠한 역할(-er ,-or)을 나타내는 것이 아닌 **어떠한 역할을 수행하는 주체** 로써 명칭을 사용하고 싶었기 때문입니다. 사실 여전히 AppState라는 이름이 객체로써 올바른 이름을 사용했는지는 잘 판단이 서질 않지만 `Module`이라는 이름은 확실히 어떤 역할을 수행하는지 전달이 잘 되지 않을 것 같다는 생각이 들었습니다.~~
  - `AppState`에서 굳이 filter라는 상태를 유지할 필요가 없다는 생각이 들었습니다. app이 실행될 때 한번 실행되고 이후에 다시 호출되는 구조가 아니라서 initAppState이라는 함수로 바꾸어 사용했습니다.

[[refactor] : handleTabButton => changeFavoriteStatus 으로 이름변경](https://github.com/woowacourse/javascript-lunch/pull/240/commits/2d7ebbbe00d4823d04ce4ab62d3f05fc03c7481c)

- `handleTabButton`
  -   `changeFavoriteStatus` 으로 이름변경. `handle`이라는 이름이 조금 모호하다라는 생각이 들었습니다. 하나의 단일 함수임에도 불구하고 `handle`이라는 이름이 붙는다면 여러개의 역할이 부여된 것이라는 생각이 들었어요. 그래서 FavoriteState을 변경한다라는 이름을 주었습니다.

<br/>

## 의문점!

> Q1) ~Handler 라는 파일 명을 짓는 것이 안좋은건가요?

해당 부분을 말씀해주시면서 `역할이 많다`라는 말을 해주셨는데 짚어주신 것 처럼 의도된 부분이거든요. 예를들어보자면 데이터와 관련된 CRUD 동작들을 함수로 작성하게 되니 여러 함수들을 묶어서 하나의 어떠한 파일로 명명하기가 애매하다는 생각이 들었어요. 그래서 'Handler'라는 명칭을 붙여 이들을 FoodItem과 관련된 함수들을 의미할 수 있도록 정의하게 되었던 것 같습니다. 

이들을 함수로 작성하게 된 이유는 `상태`를 관리할 필요가 없었기에 Class로는 사용할 이유가 없었어요. **혹시나 제가 잘못 생각하고 있는 부분이 있을까요? 그리고 위와 같이 `데이터와 관련된 CRUD 동작들을 함수로 작성하게 되`는 경우엔 어떻게 파일 명을 작성하는게 전달이 더 잘 될지 감이 오지 않아요.** 이럴땐 어떻게 이름을 작성하는 것이 좋을지에 대한 의견을 여쭤보고싶습니다. 
 +) `FoodItemService`라는 파일명도 생각해보았지만 `Service` 라는 명칭이 더 모호한 것 같다는 생각이 들어요. (제가 아직 이름 짓는 부분이 미숙해서 그런 것 같기도 합니다)

cf) 함수로 작성하게 된 과정을 설명하게 된 이유는 class로 만약 사용을 하게 되었다면 "객체"로써의 의미를 가질 수 있도록 네이밍을 조정했을 것 같아요. class를 명명할때의 저만의 기준이 있는데 `-er` 또는 `-or` 과 같이  "**어떠한 역할**"이 아니라 어떠한 역할을 해주는 "**객체**"로써의 이름을 지으려고 노력하는 편입니다.

> 한줄요약 : **상태**관리가 필요없다 -> class로 쓸 이유가 없다 -> 같은 범주로 묶이지만 각자 역할이 다른 함수들을 감싸는 파일 이름이 필요해졌다 -> ~handler로 묶으면 되겠다!


<br/>

> Q2) `isFavoriteState` 이라는 이름이 모호하다고 생각하시는 이유가 있을까요?

의도했던 바는 현재 Tab바의 상태가 `자주가는 음식점`인지 아닌지를 `boolean` 형식으로 반환하도록 했습니다. `Favorite`이라고 명칭을 칭한 이유는 즐겨찾기 버튼이라는 이름에 착안해서 나온 이름인데 혹시 이 부분이 제대로 전달이 되지 않았을까요?

<br/>

## 한줄소감..

음 함수명 정하는게 이렇게 어려운 일일줄 몰랐네요.. 아직 네이밍을 많이 해보지 않아서 그런 것 같습니다.. 말씀하신대로 "리팩터링" 책을 한번 읽어봐야겟다는 생각도 들어요. 지금 외에도 추가적으로 이름 변경이 필요해지는 경우 계속 추가해두도록 하겟습니다!

mun-kyeong committed:
[refactor] : AppState => initAppState라는 함수로 수정

- 이유 : filter라는 상태를 굳이 유지 할 필요가 없다고 생각됨.

woowa-euijinkk commented:
@mun-kyeong 
한층 더 나은 네이밍이 되었네요.
책도 읽어보고, 다른 사람은 어떻게 지었는지, 보통 어떤 식의 naming convention 을 쓰는지 조사하고, 노력하다보면 익숙해질거에요!


## handler 네이밍에 대해
handler 가 이만큼 있네요.
<img width="603" alt="image" src="https://github.com/user-attachments/assets/3c0d1593-8d48-4f23-9504-10aa383ac177" />

파일명이 handler 이고, 그 안에 그 기능을 다루는 함수를 몰아 넣으셨군요.

`domain/handler/*` 폴더 구조가 있는데요.

1. domain/foodItem 이라고만 해도, foodItem 을 handling 하는 폴더임을 알 수 있다. handler = 다루는 것 => 어떤 의미도 없다고 생각이 드네요
2. handle = "다루다" 같은 넓은 의미로는 쓰이지 않는거같아요. 엄청 모호한 이름인거죠. 쓰이는 맥락은 event handler 말고는 기억이 나지 않네요. handleClick, onClickHandler 정도로만요. 

캉골이 쓰는 모호한 단어들을 모두 걷어내고, 확실하게 해보시죠! 

캉골 GPT 많이 쓰시나요? GPT 와 변수명을 피드백 받고, 일반적으로 사용되는 이름, 더 명시적인 이름으로 지어보아요!

## isFavoriteState 네이밍에 대해

1. `isFavoriteState` 에서 state 는 아무 의미가 없습니다. `isFavorite` 라고 해도 되겠죠. 
2. is~ 는 함수가 아니라 변수에 쓰입니다. 함수는 getIsFavorite 처럼 되어야겠죠. 함수에 보통 붙는 네이밍 (get, generate, calculate, ...) 그리고 변수에 보통 붙는 네이밍 (is, should, can, ~ed) 등을 학습해보고, 컨벤션에 맞게 써보아요. 
3. isFavorite 라고 하면, 이 음식점이 좋아요 눌러졌는지 판단하는 것 같아요. 정확히는 tab 이 favorite mode 인지, favorite가 selected 되었는지를 나타내는 변수인거죠. 

`리팩터링` 과 `클린코드`를  읽다보면 감각이 더 생길거에요!

mun-kyeong commented:
>domain/foodItem 이라고만 해도, foodItem 을 handling 하는 폴더임을 알 수 있다. handler = 다루는 것 => 어떤 의미도 없다고 생각이 드네요
 
아 어떤 의미에서 Handler가 모호하다고 말씀하신지 알 것 같아요. 주변 크루들에게도 물어봤을 때 handler가 eventHandler 를 의미하는 줄 알았다고 처음에 그러더라구요. 설명을 통해서도 바로 의미가 전달이 되지 않는다면 그 **모호함을 제공하는 부분을 없애**는게 좋을 것 같다는 생각이 듭니다.

---

> 캉골 GPT 많이 쓰시나요? GPT 와 변수명을 피드백 받고, 일반적으로 사용되는 이름, 더 명시적인 이름으로 지어보아요!

GPT를 변수네이밍을 지을 때 쓰려고 하는 편인데 변수명이 길어지는 경우가 많더라구요.. 그래서 너무 긴 변수명은 사용하기가 부담스러워서 제멋대로 짓다보니 모호함이 추가된 듯 싶습니다 😂
지금 생각해보니 함수에 담긴 역할이 많아서 GPT도 수식어가 많이 붙은 이름을 추천해준 것 같아요. 이를 통해서도 함수의 역할 분리를 파악할 수 있을 것 같네요 ㅎㅎ

---

> 1. isFavoriteState 에서 state 는 아무 의미가 없습니다. isFavorite 라고 해도 되겠죠.
> 2. is~ 는 함수가 아니라 변수에 쓰입니다. 함수는 getIsFavorite 처럼 되어야겠죠. 

함수 이름에서 역할이 드러나지 않았다는게 아니라 함수 명명 방식에 대한 부분이였군요! 말씀하신대로 ~is로 시작하는 변수명들은 모두 함수가 아닌 변수로 사용했던 것 같아요. boolean 타입을 다룬다는 이유만으로 함수 네이밍을 지은 것 같다는 생각이 들어요😂 잘 설명해주셔서 감사합니다!! 
해당 부분까지 반영해서 다시 PR 올려두도록 하겠습니다. 늦어도 내일 안까지는 PR 마무리해서 올리겠습니다. 기다려주셔서 감사해요 ☺️

mun-kyeong committed:
[refactor] : handler라는 명칭 제거

- FoodStorageHandler => FoodStorage
- FoodItemHandler => Food
- TabButtonHandler => FavoriteService

mun-kyeong committed:
[refactor] : 파일명에서 ~handler 제거

mun-kyeong committed:
[refactor] : isFavoriteState 명칭 변경 => favoriteState

mun-kyeong commented:
## 이후 개선점
- 파일명에서 ~handler 제거
  - [[refactor] : handler라는 명칭 제거](https://github.com/woowacourse/javascript-lunch/pull/240/commits/1406572a256e5b1b8f929506eaabd35cedd57650)
  - [[refactor] : 파일명에서 ~handler 제거](https://github.com/woowacourse/javascript-lunch/pull/240/commits/5218f35b9c894f0ed5982d0b54fa824828e234f1)

- isFavoriteState 명칭 변경 => favoriteState
  - [[refactor] : isFavoriteState 명칭 변경 => favoriteState](https://github.com/woowacourse/javascript-lunch/pull/240/commits/5428f487edef896a792928ff510abdde5d15d900)

- filter 파라미터로 전달받는 부분 변경 (삭제)
  - [[refactorr : filter props로 전달받는 부분 제거](https://github.com/woowacourse/javascript-lunch/pull/240/commits/21b686322d0bc027bf1073ed27f27f2b23d5fd91)

- foodItems라는 명칭과 FoodList라는 명칭이 혼용되어 사용됨. => foodList로 명칭 동일
  - [[refactor] : foodItems => foodList로 명칭 동일](https://github.com/woowacourse/javascript-lunch/pull/240/commits/6432d95e36b91283f086389659df5720c0a0f149)

- 폴더구조 변경
  - class가 아닌 function 으로 이루어진경우 대문자시작 => 카멜 케이스로 변경
  - 상태 관리가 필요없는 경우 class에서 function으로 변경
  - [[refactor] : 폴더구조 변경](https://github.com/woowacourse/javascript-lunch/pull/240/commits/58db3ece4dd232f34c70a8f3417a958cd81523a0)

mun-kyeong committed:
[refactorr : filter props로 전달받는 부분 제거

mun-kyeong committed:
[refactor] : foodItems => foodList로 명칭 동일

mun-kyeong committed:
[refactor] : 폴더구조 변경

- domain 폴더 안에서 page/utils/service 구분해서 파일 정리
- class가 아닌 function 으로 이루어진경우 대문자시작 => 카멜 케이스로 변경
- types 파일 변경
- 상태 관리가 필요없는 경우 class에서 function으로 변경

mun-kyeong committed:
[refactor] : FoodListPage에서 반복되는 Filter 부분 컴포넌트화

mun-kyeong committed:
[refactor] : FavoriteFilter 싱글톤 배턴으로 변경

mun-kyeong committed:
[refactor] : boolean type의 경우 is~ 로 명칭 통일 + isFavoriteFilterActive 변수명 사용

mun-kyeong committed:
[refactor] : favoriteFilteredFoodList 명칭 변경

mun-kyeong committed:
[refactor] : 리터럴 타입으로 필터 타입 추론

mun-kyeong committed:
[refactor] : FoodType에서 category 속성 추가 & 비교

mun-kyeong committed:
[refactor] : filterAndSortFoodList로 명칭 변경 - sorting이라는 키워드 추가

mun-kyeong committed:
[refactor] :  SortedFoodList 중복 기능 제거

mun-kyeong committed:
[refactor] : SORTING_OPTIONS, CATEGORY_OPTIONS 상수화

mun-kyeong committed:
[refactor] : Filter 싱글톤 배턴으로 변경

mun-kyeong commented:
기다려주셔서 감사합니다 시지프 🥹🥹
조금 오래 걸리긴 했지만 지금 다시 코드를 보니 저조자도 어떤 함수인지 눌러봐야 이해되는 함수명들이 있었던 것 같아요 

시지프가 말해준 것 처럼 다음 미션에서부턴 함수명만을 보고도 어떻게 동작하는지 이해할 수 있도록 코드를 작성해보려합니다. 제게 문제점이 무엇인지를 파악할 수 있어서 어떻게 공부해야할지도 감이 올 것 같아요. 

시지프가 추천해준 리팩터링이나 클린코드 책도 꼭 읽어보고 블로그로 후기 남기겠습니다 😊😊

추가적으로 수정이 필요한 부분이 있다면 알려주세요! 더 나은 코드를, 이해하기 쉬운 코드를 작성하고싶습니다. 
평소에 제가 가지고 있던 답답함들을 깨달을 수 있었던 미션이라 제게 의미가 많을 것 같아요. 다시한번 감사합니다 시지프 😊😊

woowa-euijinkk commented on src/domain/page/loadFoodListPage.ts:
```diff
@@ -0,0 +1,69 @@
+  container.innerHTML = `
+      <ul class="restaurant-list">
+      </ul>
+    `;
```
이런 부분들이 다 함수로 분리되어야 하지 않을까요?
주석 처리가 아니라!

woowa-euijinkk commented on src/domain/FavoriteFilter.ts:
```diff
@@ -0,0 +1,47 @@
+    updateFoodList({ foodItem });
+    showConvertedItem({ showOnlyFavorites: this.currentStatus() });
+    event.stopPropagation();
+  }
```
결합도가 매우 높습니다.
favoriteFilter 가 updateFoodList, newFoodItem, showConvertedItem 를 모두 알아야 하고, 
갑작스런 event.stopPropagation(); 맥락까지 있네요.

그냥 여러 기능을 섞어놓은 함수입니다. 유지보수가 매우 어려운 코드입니다.

이게 근본적으로 FavoriteFilter 의 역할이 모호하기 때문에 발생한다고 생각해요.
FavoriteFilter 가 무엇인가요? 
이름만 봐서는 자주 가는 음식점, 을 필터링 해주는 것? 으로 생각되는데요.
그 기능이 아닌 것 같습니다.
그러니까, `모든 음식점` , `자주 가는 음식점` 필터하는 것 이라고 생각하지만, toggleStatus 는 무엇을 하고 있나요?
foodItem 의 favorite 를 바꾸고 있습니다.

FavoriteFilter 이름이 모호하고, 그 내부 상태 isFavorite 도 모호합니다. 제가 이전에 isFavoriteMode 로 리팩터링 하길 권했는데, 이 이름이 헷갈리기 때문에, 혼동이 계속 오는 것입니다.

"함수/클래스의 역할을 한줄로 정의하고, 하나의 함수명을 부여해보는 것" 을 계속 연습해보길 바라요.

woowa-euijinkk commented on src/domain/FavoriteFilter.ts:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  toggleStatus({ event, foodItem }: ToggleStatusType) {
+    console.log(this.isFavorite);
```
하나하나 대충대충 빠르게 하는 것 보다, 완벽하게 해내는 습관이 필요합니다.
console 지우고, 불필요한 주석도 다 지워보아요!

캉골에겐 이런 디테일이 필요하기 때문에, 사소한 코멘트도 남깁니다.

woowa-euijinkk commented on src/domain/SortingFilter.ts:
```diff
@@ -3,9 +3,9 @@ import {
+} from "../types/domain/SortingFilterType";
 
-export class Filter {
+export class SortingFilter {
```
Sort 는 정렬이고,
Filter 는 깔대기 입니다.

SortingFilter 는 어떤 의미인가요? 요런 이름도 하나하나 신경써보아요

woowa-euijinkk commented on src/domain/SortingFilter.ts:
```diff
@@ -63,3 +71,5 @@ export class SortingFilter {
   }
 }
+
+export const sortingFilter = SortingFilter.getInstance();
```
요런거 싱글톤은 좋네요!

woowa-euijinkk replied:
한번만 실행되면 되는 기능을 개발할 때, 함수로 할 수도 있고, 싱글톤으로 해볼 수 있겠네요! 
잘 선택하신 것 같아요 👍

woowa-euijinkk reviewed (approved):
머지하기엔 부족한 점이 많지만, 다음에도 학습 기회가 많으므로, 이만 머지합니다.
저의 코멘트들 꼭 잊지 말고, 유지해보아요. 궁금한거 있으면 더 질문해주셔도 돼요
꼼꼼하고, 신뢰할 만한 코드를 짜고, 일관성 있고, 불필요한 추상화 하지 않는, 그런 코드를 짜보아요.
"대충 짜기" 금지!

많은거 수정하고, 고민하느라 고생 많으셨어요! 화이팅!
